src/utils/flow.js

```diff
@@ -33,33 +42,62 @@ function convertFlowObjectTypeAnnotation(
   }, {});
 }
 
+function convertFunctionTypeAnnotation(
+  objectType: FunctionTypeAnnotation,
+  flowTypes: { [name: string ]: ObjectTypeAnnotation },
+): FlowTypes {
+  return objectType.params.reduce((obj, param) => {
+    const key = param.name.name;
+
+    return {
+      ...obj,
+      [key]: convertTypeAnnotationToFlowType(param.typeAnnotation, param.optional, flowTypes)
+    };
+  }, {});
+}
+
 export function convertTypeAnnotationToFlowType(
   value: TypeTypeAnnotation,
   nullable: boolean,
-  flowTypes: { [name: string ]: ObjectTypeAnnotation }
+  flowTypes: { [name: string ]: ObjectTypeAnnotation },
+  variables?: Object
 ): FlowType {
   if (value.type === "NullableTypeAnnotation") {
-    return convertTypeAnnotationToFlowType(value.typeAnnotation, true, flowTypes);
+    return convertTypeAnnotationToFlowType(value.typeAnnotation, true, flowTypes, variables);
   }
 
-  if (value.type === "GenericTypeAnnotation" && flowTypes[value.id.name]) {
-    return convertTypeAnnotationToFlowType(flowTypes[value.id.name], nullable, flowTypes);
+  if (value.type === "GenericTypeAnnotation") {
+    if (flowTypes[value.id.name]) {
+      return convertTypeAnnotationToFlowType(flowTypes[value.id.name], nullable, flowTypes, variables);
+    }
+
+    if (value.typeParameters) {
+      if (value.id.name === "PropertyWithArgs") {
+        const prop = value.typeParameters.params[0];
+
+        return convertTypeAnnotationToFlowType(prop.returnType, nullable, flowTypes, convertFunctionTypeAnnotation(prop, flowTypes));
+      }
+
+      if (value.id.name === "PropertyVariable") {
+        return convertTypeAnnotationToFlowType(value.typeParameters.params[0], nullable, flowTypes, variables);
+      }
+    }
   }
 
   if (value.type === "ObjectTypeAnnotation") {
     return {
       type: "object",
       nullable,
-      properties: convertFlowObjectTypeAnnotation(value, flowTypes)
+      properties: convertFlowObjectTypeAnnotation(value, flowTypes, variables)
```

Comment:
Maybe we should follow the GraphQL schema and change this to fields and put variables on a field instead?